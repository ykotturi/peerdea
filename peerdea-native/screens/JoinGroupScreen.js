import React from "react";
import {
  Image,
  Platform,
  ScrollView,
  StyleSheet,
  Text,
  TouchableOpacity,
  TouchableHighlight,
  View,
  Alert,
  TextInput,
  Dimensions,
  Keyboard,
  KeyboardAvoidingView,
} from "react-native";
// import { WebBrowser } from "expo";
// import { Button } from "react-native-elements";
import AsyncStorage from "@react-native-async-storage/async-storage";
// import { KeyboardAwareScrollView } from "react-native-keyboard-aware-scroll-view";
import Sha256 from "../components/sha256.js";
import styles from "../components/style";
const { width, height } = Dimensions.get("window");
import {DOCKER_URL} from "@env"

export default class JoinGroupScreen extends React.Component {
  state = {
    groupName: "",
    groupPassword: "",
  };

  async onJoin() {
    try {
      var groupName = this.state.groupName;
      var values = await AsyncStorage.multiGet(["username"]);
      var username = values[0][1];
      var values = await AsyncStorage.multiGet(["username"]);
      var username = values[0][1];
      var query = `query{group(name: "${groupName}"){name password id}}`;
      //check if the group exists first
      const checkRes = await fetch(
        `http://${DOCKER_URL}/graphql?query=` + query,
        { method: "GET" }
      );
      const checkResJson = await checkRes.json();
      var groupID;
      if (checkResJson.data.group != null)
        var groupID = checkResJson.data.group.id;
      //if the group does not exist, notify the user to create a new group name
      if (checkResJson.data.group == null) {
        //new error checking for graphQL
        Alert.alert(
          "Group " + groupName + " does not exist",
          "Please try again with a different group name",
          [{ text: "OK", onPress: () => console.log("OK Pressed") }],
          { cancelable: false }
        );
      }
      // //if the group password is incorrect, notify the user
      // else if (
      //   unhashed != checkResJson.data.group.password &&
      //   checkResJson.data.group.password != null &&
      //   groupPassword != checkResJson.data.group.password
      // ) {
      //   Alert.alert("Your group password is incorrect.");
      // }
      //if the group exists, redirect the screen to the create concept screen
      else {
        await AsyncStorage.multiSet([
          ["groupName", groupName],
          ["groupID", groupID],
        ]);

        //add user to list of users in the group
        userList = `query{group(name: "${groupName}"){users}}`;
        const checkResUserList = await fetch(
          `http://${DOCKER_URL}/graphql?query=` + userList,
          { method: "GET" }
        );
        const checkResUserListJson = await checkResUserList.json();
        //decreased to 5 for debugging, return to 25 for live deployments
        if (
          checkResUserListJson.data.group.users.length >= 25 &&
          !checkResUserListJson.data.group.users.includes(username)
        ) {
          Alert.alert(
            "This group is already full! Try creating or joining another group."
          );
          return;
        } else if (
          checkResUserListJson.data.group.users.includes(username) == false
        ) {
          var addUser = `
          mutation {
            addUserToGroup(username: "${username}", groupName: "${groupName}"){
              users
            }
          }`;
          const checkAddUser = await fetch(
            `http://${DOCKER_URL}/graphql?query=` + addUser,
            { method: "POST" }
          );
          const checkAddUserJson = await checkAddUser.json();

          //check if the group is already in the user's groups record
          query = `query{user(username: "${username}"){groups}}`;
          const checkRes = await fetch(
            `http://${DOCKER_URL}/graphql?query=` + query,
            { method: "GET" }
          );
          const checkResJson = await checkRes.json();

          //if the group is not in the groups record, add it
          if (
            checkResJson.data.user.groups == null ||
            checkResJson.data.user.groups.includes(groupName) == false
          ) {
            query = `
            mutation{
              addGroupToUser(username: "${username}", groupName: "${groupName}"){
                groups
              }
            }`;
            const checkAddGroup = await fetch(
              `http://${DOCKER_URL}/graphql?query=` + query,
              { method: "POST" }
            );
            const checkAddGroupJson = await checkAddGroup.json();
          }
        }

        this.props.navigation.navigate("Third");
      }
    } catch (err) {
      console.log(err);
    }
  }

  render() {
    return (
      <View style={{ flex: 1 }}>
        <ScrollView
          keyboardShouldPersistTaps="handled"
          contentContainerStyle={{
            flexGrow: 1,
            justifyContent: "center",
            alignItems: "center",
          }}
          style={{ width: "100%", backgroundColor: "#fff" }}
        >
          <KeyboardAvoidingView enabled>
            <View style={styles.container}>
              <Image
                style={{ width: 300, height: 51 }}
                source={require("../assets/images/peerdea-logo-draft.png")}
                accessible={true}
                accessibilityLabel="Peerdea app logo"
              />
              <Text style={styles.getStartedText}>Join a group below:</Text>

              <TextInput
                accessibilityLabel="groupName"
                style={styles.hintText}
                onChangeText={(text) =>
                  this.setState({ groupName: text.trim() })
                }
                placeholder="Enter your group name"
                //workaround to prevent autofill
                textContentType="oneTimeCode"
                ref={(ref) => {
                  this._groupNameinput = ref;
                }}
                returnKeyType="next"
                onSubmitEditing={() =>
                  this._groupPasswordinput && this._groupPasswordinput.focus()
                }
                blurOnSubmit={false}
              />

              {/* <TextInput
                accessibilityLabel="groupPassword"
                style={styles.hintText}
                onChangeText={(text) => this.setState({ groupPassword: text })} //don't trim the text, as passwords can end in whitespace
                placeholder="Enter your group password"
                secureTextEntry={true}
                //workaround to prevent autofill
                textContentType="oneTimeCode"
                ref={(ref) => {
                  this._groupPasswordinput = ref;
                }}
                returnKeyType="done"
                onSubmitEditing={Keyboard.dismiss}
                blurOnSubmit={false}
              /> */}

              <TouchableHighlight
                style={styles.buttonLargeLeft}
                onPress={() => this.onJoin()}
              >
                <Text style={styles.buttonText}>Join group</Text>
              </TouchableHighlight>
            </View>
          </KeyboardAvoidingView>
        </ScrollView>
      </View>
    );
  }
}
