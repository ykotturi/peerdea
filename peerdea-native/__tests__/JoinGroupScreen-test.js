import React from 'react';
import renderer from 'react-test-renderer';

import JoinGroupScreen from '../screens/JoinGroupScreen';

jest.mock('@react-native-async-storage/async-storage', () =>
  require('@react-native-async-storage/async-storage/jest/async-storage-mock')
);

describe('Join group Screen', () => {
  jest.useFakeTimers();

  it(`renders the join group screen`, () => {
    const mockNavigation = {
      navigate: jest.fn(),
      addListener: jest.fn().mockImplementation((event, callback) => {
        callback();
        //returning value for `navigationSubscription`
        return { remove: jest.fn() }
      }),
    };

    const tree = renderer.create(
      <JoinGroupScreen navigation={mockNavigation} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
