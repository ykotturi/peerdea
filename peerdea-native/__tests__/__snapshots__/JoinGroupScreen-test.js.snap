// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Join group Screen renders the join group screen 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "alignItems": "center",
        "flexGrow": 1,
        "justifyContent": "center",
      }
    }
    keyboardShouldPersistTaps="handled"
    style={
      Object {
        "backgroundColor": "#fff",
        "width": "100%",
      }
    }
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "flex": 1,
              "justifyContent": "center",
              "paddingBottom": 30,
              "paddingTop": 18,
            }
          }
        >
          <Image
            accessibilityLabel="Peerdea app logo"
            accessible={true}
            source={1}
            style={
              Object {
                "height": 51,
                "width": 300,
              }
            }
          />
          <Text
            style={
              Object {
                "color": "rgba(96,100,109, 1)",
                "fontSize": 17,
                "lineHeight": 20,
                "marginBottom": 10,
                "marginLeft": 10,
                "marginRight": 10,
                "textAlign": "center",
              }
            }
          >
            Join a group below:
          </Text>
          <TextInput
            accessibilityLabel="groupName"
            blurOnSubmit={false}
            onChangeText={[Function]}
            onSubmitEditing={[Function]}
            placeholder="Enter your group name"
            returnKeyType="next"
            style={
              Object {
                "borderColor": "gray",
                "borderRadius": 5,
                "borderWidth": 1,
                "height": 40,
                "marginBottom": 10,
                "paddingHorizontal": 3,
                "width": "100%",
              }
            }
            textContentType="oneTimeCode"
          />
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "#0076ff",
                "borderColor": "#0076ff",
                "borderRadius": 9,
                "borderWidth": 1,
                "elevation": 3,
                "height": 50,
                "justifyContent": "center",
                "width": "90%",
              }
            }
          >
            <Text
              style={
                Object {
                  "alignItems": "center",
                  "color": "#ffffff",
                  "fontSize": 20,
                  "justifyContent": "center",
                  "paddingHorizontal": "8%",
                }
              }
            >
              Join group
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
